tools/create_snapshot_artifact/main.py

```diff
@@ -0,0 +1,339 @@
+
+    root_path = tempfile.mkdtemp(
+        prefix=MicrovmBuilder.ROOT_PREFIX,
+        dir=f"{DEFAULT_TEST_SESSION_ROOT_PATH}")
+
+    # Compile new-pid cloner helper.
+    bin_cloner_path = compile_file(
+        file_name='newpid_cloner.c',
+        bin_name='newpid_cloner',
+        dest_path=root_path
+    )
+    # Compile vsock helper.
+    bin_vsock_path = compile_file(
+        file_name='vsock_helper.c',
+        bin_name='vsock_helper',
+        dest_path=root_path
+    )
+
+    # Fetch kernel and rootfs artifacts from S3 bucket.
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    kernel_artifacts = ArtifactSet(artifacts.kernels())
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword="ubuntu"))
+
+    # Create a test context.
+    test_context = TestContext()
+    test_context.custom = {
+        'network_config': net_config,
+        'bin_cloner_path': bin_cloner_path,
+        'session_root_path': root_path,
+        'bin_vsock_path': bin_vsock_path
+    }
+
+    # Create the test matrix.
+    test_matrix = TestMatrix(context=test_context,
+                             artifact_sets=[
+                                 kernel_artifacts,
+                                 disk_artifacts
+                             ])
+
+    test_matrix.run_test(create_snapshots)
+
+
+def create_snapshots(context):
+    """Snapshot microVM built from vm configuration file."""
+    vm = setup_vm(context)
+
+    # Get ssh key from read-only artifact.
+    ssh_key = context.disk.ssh_key()
+    ssh_key.download(vm.path)
+    vm.ssh_config['ssh_key_path'] = ssh_key.local_path()
+    os.chmod(vm.ssh_config['ssh_key_path'], 0o400)
+
+    _configure_vm_from_json(vm, VM_CONFIG_FILE)
+    configure_network_interfaces(vm)
+    vm.spawn()
+
+    # Ensure the microVM has started.
+    response = vm.machine_cfg.get()
+    assert vm.api_session.is_status_ok(response.status_code)
+    assert vm.state == "Running"
+
+    # Populate MMDS.
+    data_store = {
+        'latest': {
+            'meta-data': {
+                'ami-id': 'ami-12345678',
+                'reservation-id': 'r-fea54097',
+                'local-hostname': 'ip-10-251-50-12.ec2.internal',
+                'public-hostname': 'ec2-203-0-113-25.compute-1.amazonaws.com'
+            }
+        }
+    }
+    populate_mmds(vm, data_store)
+
+    # Iterate and validate connectivity on all ifaces after boot.
+    for iface in net_ifaces:
+        vm.ssh_config['hostname'] = iface.guest_ip
+        ssh_connection = net_tools.SSHConnection(vm.ssh_config)
+        exit_code, _, _ = ssh_connection.execute_command("sync")
+        assert exit_code == 0
+
+    # Validate MMDS.
+    validate_mmds(ssh_connection, data_store)
+
+    root_path = context.custom['session_root_path']
+    vsock_helper = context.custom['bin_vsock_path']
+    vm_blob_path = "/tmp/vsock/test.blob"
+    # Generate a random data file for vsock.
+    blob_path, blob_hash = make_blob(root_path)
```

Comment:
not sure why we do this before creating a snapshot.
I believe this needs to be done after resuming the snapshot (in the upcoming test), since it is part of the process of validating that the vsock device works

Reply:
It is, indeed, part of the process of validating that the vsock device works. However, here we copy the blob to the guest, so in the upcoming test we will be able to use it to validate vsock device. This way, we also ensure that the file is persistent into the guest's memory. If you don't think that it brings enough value, I would be happy to discuss this.

Reply:
Personally I find it confusing. It's a bit weird to see this as a snapshot artifact and honestly I believe this is best suited in the upcoming test. I think we should generate this random blob and copy it in the VM after it's resumed

Reply:
We do this in other tests, too, but I see your point that it's confusing to have the blob hash as a snapshot artifact. I moved the blob generation to the device validation test.